File: cdm/cpp/properties/SEScalarPower.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarPower.h"
#include "bind/cdm/Properties.pb.h"

const PowerUnit PowerUnit::W("W");
const PowerUnit PowerUnit::kcal_Per_s("kcal/s");
const PowerUnit PowerUnit::kcal_Per_hr("kcal/hr");
const PowerUnit PowerUnit::kcal_Per_day("kcal/day");
const PowerUnit PowerUnit::J_Per_s("J/s");
const PowerUnit PowerUnit::BTU_Per_hr("BTU/hr");

bool PowerUnit::IsValidUnit(const std::string& unit)
{
  if (W.GetString().compare(unit) == 0)
    return true;
  if (kcal_Per_s.GetString().compare(unit) == 0)
    return true;
  if (kcal_Per_hr.GetString().compare(unit) == 0)
    return true;
  if (kcal_Per_day.GetString().compare(unit) == 0)
    return true;
  if (J_Per_s.GetString().compare(unit) == 0)
    return true;
  if (BTU_Per_hr.GetString().compare(unit) == 0)
    return true;
  return false;
}

const PowerUnit& PowerUnit::GetCompoundUnit(const std::string& unit)
{
  if (W.GetString().compare(unit) == 0)
    return W;
  if (kcal_Per_s.GetString().compare(unit) == 0)
    return kcal_Per_s;
  if (kcal_Per_hr.GetString().compare(unit) == 0)
    return kcal_Per_hr;
  if (kcal_Per_day.GetString().compare(unit) == 0)
    return kcal_Per_day;
  if (J_Per_s.GetString().compare(unit) == 0)
    return J_Per_s;
  if (BTU_Per_hr.GetString().compare(unit) == 0)
    return BTU_Per_hr;
  std::stringstream err;
  err << unit << " is not a valid Power unit";
  throw CommonDataModelException(err.str());
}

void SEScalarPower::Load(const cdm::ScalarPowerData& src, SEScalarPower& dst)
{
  SEScalarPower::Serialize(src, dst);
}
void SEScalarPower::Serialize(const cdm::ScalarPowerData& src, SEScalarPower& dst)
{
  SEUnitScalar::Serialize(src.scalarpower(), dst);
}

cdm::ScalarPowerData* SEScalarPower::Unload(const SEScalarPower& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarPowerData* dst = new cdm::ScalarPowerData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarPower::Serialize(const SEScalarPower& src, cdm::ScalarPowerData& dst)
{
  SEUnitScalar::Serialize(src, *dst.mutable_scalarpower());
}

File: src/java/pulse/cdm/properties/SEScalarPower.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.properties;

import pulse.cdm.bind.Properties.ScalarPowerData;

import pulse.cdm.properties.CommonUnits.PowerUnit;

/**
 * @author abray
 * Power Scalar Class, will enforce that units are proper to Powers
 */

public class SEScalarPower extends SEScalar
{
  public SEScalarPower()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarPower(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarPower(double value, PowerUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarPowerData src, SEScalarPower dst)
  {
    SEScalar.load(src.getScalarPower(),dst);
  }
  public static ScalarPowerData unload(SEScalarPower src)
  {
    if(!src.isValid())
      return null;
    ScalarPowerData.Builder dst = ScalarPowerData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarPower src, ScalarPowerData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarPowerBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, PowerUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(PowerUnit unit)
  {
    return this.getValue(unit.toString());
  }

  public boolean validUnit(String unit)
  {
    if(PowerUnit.validUnit(unit))
      return true;
    return false;
  }
}
